### ISSUE TITLE
spec(macro): ngFactory state

### ISSUE THREAD

**Issue by douglasduteil** [douglasduteil/angular-utility-belt#12]
Hi guys let's briefly say where we at for me.
# NgFactory for development
- [x] test the sources with karma.
- [x] test coverage info
- [x] code quality with jshint
- [ ] Dev server serving the example files
- [ ] flexible js and CSS preprocessors ( allow none or es6 or browerify ... and sass or less or stylus...)
# NgFactory for distribution
### Markdown version
- [x] Generate the API doc in markdown
- [ ] Inline example files in the markdown
### HTML version
- [ ] Generate the API doc in html
- [ ] Run various examples 
## Publication

Following #11
- [ ] Have a bump system
- [ ] Have a auto git push to dist branch system
- [ ] Have a changelog generation

I might forgot stuff. Don't hesitate to add it.
We still have a lot to do.

**Comment by revolunet** [comment]
thanks

about "inline example in markdown" : you mean embedding the raw *.html and js from docs/examples folder ?

**Comment by douglasduteil** [comment]
> about "inline example in markdown" : you mean embedding the raw *.html and js from docs/examples folder ?

Yep

**Comment by mgcrea** [comment]
@douglasduteil thanks for the sum up.

On my side, I'm pretty unhappy about the state of the task system, I'm going to try to refactor things (this evening) to keep it as simple as possible. I'm still unsure about how transforms should be linked to your project.

I think having some meta-package to handle theses transforms stuff might be a good thing, just to clean things up, and abstract this mess away. But there is some hard questions that must be answered:
1. Having the said abstract `gulp-tranforms` package hard require every dep (`uglify`, `jade`, `ngmin`, etc.) would probably be an anti-pattern, as the initial objective was to decorrelate from empty gulp-plugin shells, not to have a single point of failure. Would also be against the point [gulp-guidelines/#2](https://github.com/gulpjs/gulp/blob/master/docs/writing-a-plugin/guidelines.md).
2. Having the `gulp-transforms` not hard linking deps, but only adding stream capabilities to existing plugins, would ideally work like bluebird's mechanics to promisify a lib (eg. `var fs = bluebird.promisifyAll(require('fs'));`). That was my initial goal: letting the end-user require `less`, `ng-annotate`, etc. directly, as-if they had the "streamify" support into the core libs. The issue is that there is no way to make this generic, and the upcoming solution would have an awkward API at best. You could also have version issues down the line (major API changes breaking the tranformify), etc.

I'll work on that this evening, trying to come up with something smart / simple. I'll let you know.

**Comment by douglasduteil** [comment]
@mgcrea 

> 1. Having the said abstract gulp-tranforms package hard require every dep (uglify, jade, ngmin, etc.) would probably be an anti-pattern

Agreed. It's not a good solution. I don't might having all those transform into the project. I just think that it's would be better to abstract individual parts. So we can focus on the macro access of the tasks.

> On my side, I'm pretty unhappy about the state of the task system, I'm going to try to refactor things

I also try 2 different refacto cause it was a mess... It's might be a good idea to extract more doc functionality so we don't need extra big transforms or tasks.

**Comment by mgcrea** [comment]
So I temporary left the transform stuff away to fix the (html) docs generation and serve. It should be working now. I also renamed a few things and reorganized stuff. Tried to simplify things.

Basically I went for an extending setup using the [docs](https://github.com/douglasduteil/angular-utility-belt/tree/master/docs) folder. It contains the nunjucks template for both the readme and the html index, and a bower.json related to the docs (hence solving the bower docs deps issue). With this setup, you can easily override blocks (eg. logo in readme). Quite like the flexibility this gives, for now theses files are mandatory but we could handle the case if they were missing, since this will be also solved by the generator (creating theses files for you), it should be OK for now.

So, `gulp serve` is aliased to `gulp docs` and it enables us to work on the templates, but IMO it could be the predefined way to hack on your project's examples. Tell me if it's working for you.

**Comment by mgcrea** [comment]
Hey @douglasduteil, I'd like to move/refactor the repo today to its final location. Afterwhat, I'll try to merge my daily workflow changes (tasks leveraging [reusable channels built with lazypipe](https://github.com/gulpjs/gulp/blob/master/docs/recipes/sharing-streams-with-stream-factories.md)). Idea is to be able to use ngFactory to use for both component and webapp worflow, I'm using the later one on a daily basis so I think it's really important to find synergy between the two workflows so that one can benefit the other.

**Comment by mgcrea** [comment]
Move started to https://github.com/angular-widgets/ng-factory/commits/master

**Comment by revolunet** [comment]
awesome ! :boom:

**Comment by mgcrea** [comment]
Example of gulp-channels reusability:
- It's a `index:dist` workflow powered by gulp-inject that reuse existing channels exported with lazypipes:

https://github.com/mgcrea/gulp-channels/blob/master/lib/channels/index/dist.js

**Comment by mgcrea** [comment]
Example of final `gulpfile.js` for the end-user:

``` js
'use strict';

var gulp = require('gulp');
var gutil = require('gulp-util');
var pkg = require('./package.json');

// CONFIG
//
var config = {
  pkg: pkg,
  type: 'app',
  module: 'com.angular.MyApp'
};

require('ng-factory')(gulp, config);

// TASKS
//
var src = config.src;
gulp.task('copy:dist', function() {
  gulp.src(['bower_components/jsoneditor/jsoneditor.min.css'], {cwd: src.cwd, base: src.cwd})
    .pipe(gulp.dest(src.dest));
});

// ALIASES
//
var runSequence = require('run-sequence');
gulp.task('default', ['build']);
gulp.task('dist', ['build']);
gulp.task('test', function() {
  runSequence('ng-factory:test/clean', 'ng-factory:src/karma~init', 'ng-factory:src/karma');
});
gulp.task('build', function() {
  runSequence('ng-factory:dist/clean', 'ng-factory:src/views', ['ng-factory:dist/views', 'ng-factory:dist/copy'], 'copy:dist');
});
gulp.task('serve', function() {
  runSequence('ng-factory:src/clean', 'ng-factory:src/views', ['ng-factory:src/serve', 'ng-factory:src/watch']);
});
```

**Comment by douglasduteil** [comment]
:+1:

**Comment by douglasduteil** [comment]
Things getting REAL :shipit:
